src/ray/common/ray_syncer/ray_syncer-inl.h

```diff
@@ -136,58 +150,123 @@ class NodeSyncConnection {
+      : RaySyncerBidiReactorBase(io_context,
+                                 remote_node_id,
+                                 std::move(message_processor),
+                                 std::move(cleanup_cb)) {}
+
+  using T::StartRead;
+  using T::StartWrite;
+
+  void Send(std::shared_ptr<const RaySyncMessage> message, bool flush) override {
+    sending_message_ = std::move(message);
+    grpc::WriteOptions opts;
+    if (flush) {
+      opts.clear_buffer_hint();
+    } else {
+      opts.set_buffer_hint();
+    }
+    RAY_LOG(DEBUG) << "[BidiReactor] Sending message to "
+                   << NodeID::FromBinary(GetRemoteNodeID()) << " about node "
+                   << NodeID::FromBinary(sending_message_->node_id());
+    StartWrite(sending_message_.get(), opts);
+  }
 
-  ~ServerSyncConnection() override;
+  void OnWriteDone(bool ok) override {
+    if (ok) {
+      io_context_.dispatch([this]() { SendNext(); }, "");
+    } else {
+      RAY_LOG(ERROR) << "Failed to send the message to: "
+                     << NodeID::FromBinary(GetRemoteNodeID());
+    }
+  }
 
-  void HandleLongPollingRequest(grpc::ServerUnaryReactor *reactor,
-                                RaySyncMessages *response);
+  void OnReadDone(bool ok) override {
+    if (!ok) {
+      return;
+    }
+    io_context_.dispatch(
+        [this, msg = std::move(receiving_message_)]() mutable {
+          RAY_CHECK(!msg->node_id().empty());
+          ReceiveUpdate(std::move(msg));
+          StartPull();
+        },
+        "");
+  }
 
  protected:
-  /// Send the message from the pending queue to the target node.
-  /// It'll send nothing unless there is a long-polling request.
-  /// TODO (iycheng): Unify the sending algorithm when we migrate to gRPC streaming
-  void DoSend() override;
-
-  /// These two fields are RPC related. When the server got long-polling requests,
-  /// these two fields will be set so that it can be used to send message.
-  /// After the message being sent, these two fields will be set to be empty again.
-  /// When the periodical timer wake up, it'll check whether these two fields are set
-  /// and it'll only send data when these are set.
-  std::vector<RaySyncMessages *> responses_;
-  std::vector<grpc::ServerUnaryReactor *> unary_reactors_;
+  void StartPull() {
+    receiving_message_ = std::make_shared<RaySyncMessage>();
+    RAY_LOG(DEBUG) << "Start reading: " << NodeID::FromBinary(GetRemoteNodeID());
+    StartRead(receiving_message_.get());
+  }
+
+  /// grpc requests for sending and receiving
+  std::shared_ptr<const RaySyncMessage> sending_message_;
+  std::shared_ptr<RaySyncMessage> receiving_message_;
 };
 
-/// SyncConnection for gRPC client side. It has customized logic for sending.
-class ClientSyncConnection : public NodeSyncConnection {
+/// Reactor for gRPC server side. It has customized logic for sending.
+class RayServerBidiReactor : public BidiReactor<ServerBidiReactor> {
```

Comment:
one confusion reading the code is hard to get a full picture of the public APIs of `RayServerBidiReactor`, which are spread through `RayServerBidiReactor`, `BidiReactor` and `RaySyncerBidiReactorBase`.

Reply:
What can't be improved is the gRPC related API and it's define [here](https://github.com/grpc/proposal/blob/master/L67-cpp-callback-api.md). I think it's like how to use the library. We probably need some wrapper later on this one.

For RayServerBidiReactor and RayServerBidiReactor, it only use `cleanup_cb_` which I can try to avoid. Once that's removed, it's basically nothing you need from RaySyncerBidiReactorBase to understand these two (good isolation).  But still, you need to understand gRPC callback API first.

I think the reason you feel this is hard to following is because the abstraction is not documented well.

I'll add comments for these class and mention their relationship there. I think following that, it should make the reviewer's life easier.